core/src/main/java/org/apache/iceberg/io/ContentCache.java

```diff
@@ -0,0 +1,310 @@
+        .add("maxTotalBytes", maxTotalBytes)
+        .add("cacheStats", cache.stats())
+        .toString();
+  }
+
+  private static class CacheEntry {
+    private final long length;
+    private final List<ByteBuffer> buffers;
+
+    private CacheEntry(long length, List<ByteBuffer> buffers) {
+      this.length = length;
+      this.buffers = buffers;
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final FileIO io;
+    private final String location;
+    private final long length;
+    private InputFile fallbackInputFile = null;
+
+    private CachingInputFile(ContentCache cache, FileIO io, String location, long length) {
+      this.contentCache = cache;
+      this.io = io;
+      this.location = location;
+      this.length = length;
+    }
+
+    private InputFile wrappedInputFile() {
+      if (fallbackInputFile == null) {
+        fallbackInputFile = io.newInputFile(location, length);
+      }
+      return fallbackInputFile;
+    }
+
+    @Override
+    public long getLength() {
+      CacheEntry buf = contentCache.getIfPresent(location);
+      if (buf != null) {
+        return buf.length;
+      } else if (fallbackInputFile != null) {
+        return fallbackInputFile.getLength();
+      } else {
+        return length;
+      }
+    }
+
+    /**
+     * Opens a new {@link SeekableInputStream} for the underlying data file, either through cache or
+     * through the inner FileIO.
+     *
+     * <p>If data file is not cached yet, and it can fit in the cache, the file content will be
+     * cached first before returning a {@link ByteBufferInputStream}. Otherwise, return a new
+     * SeekableInputStream from the inner FIleIO.
+     *
+     * @return a {@link ByteBufferInputStream} if file exist in the cache or can fit in the cache.
+     *     Otherwise, return a new SeekableInputStream from the inner FIleIO.
+     */
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        // read-through cache if file length is less than or equal to maximum length allowed to
+        // cache.
+        if (getLength() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        // fallback to non-caching input stream.
+        return wrappedInputFile().newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(
+            e, "Failed to open input stream for file %s: %s", location, e.toString());
+      } catch (IOException e) {
+        throw new RuntimeIOException(
+            e, "Failed to open input stream for file %s: %s", location, e.toString());
+      }
+    }
+
+    @Override
+    public String location() {
+      return location;
+    }
+
+    @Override
+    public boolean exists() {
+      CacheEntry buf = contentCache.getIfPresent(location);
+      return buf != null || wrappedInputFile().exists();
+    }
+
+    private CacheEntry cacheEntry() {
```

Comment:
Something to think about (possibly create an issue), but this isn't really read-through, but rather read-ahead since we're reading the whole file before allowing the reader to process the manifest file.  We could improve this by making it a read-through buffer.

Reply:
Filed #5814 as a follow up.